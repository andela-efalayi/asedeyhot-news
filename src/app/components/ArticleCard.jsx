import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Card, CardActions, CardHeader,
  CardMedia, CardText } from 'material-ui/Card';
import swal from 'sweetalert2';
import IconButton from 'material-ui/IconButton';
import FavoriteBorder from 'material-ui/svg-icons/action/favorite-border';
import Favorite from 'material-ui/svg-icons/action/favorite';
import AppFavourites from '../actions/AppFavourites';

/**
 * @class ArticleCard
 * @extends {Component}
 * @method saveToFavouriteArticles
 * @param {object} props
 */

class ArticleCard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: this.props.article,
      favIcon: <FavoriteBorder />,
      tooltip: 'Add to favourite articles'
    };
    this.saveToFavouriteArticles = this.saveToFavouriteArticles.bind(this);
  }

  /**
   * Save a user's favourite article to the database
   * @memberof ArticleCard
   */

  saveToFavouriteArticles() {
    const article = this.props.article;
    AppFavourites.saveToFavouriteArticles(article)
    .then((status) => {
    /**
     * Update view if successful
     */
      this.setState({
        favIcon: <Favorite />,
        tooltip: status
      });
    })
    .catch((error) => {
    /**
     * Show an alert if unsuccessful
     */
      swal(
        'Sorry',
         error,
        'error'
      );
    });
  }

  /**
   * @param {string} word
   * @param {number} limit
   * @param {string} endChar
   * @return {string} newText
   * @memberof ArticleCard
   */

  setChar(word, limit, endChar) {
    const newText = word;
    if (newText === null) {
      return 'No text available';
    }
    return `${newText.substring(0, limit) + endChar}`;
  }
  render() {
    return (
      <Card className="card with-image">
      <CardHeader className="article-title"
      title={this.setChar(this.props.article.title, 36, '...')} />
      <CardMedia>
        <img src={this.props.article.urlToImage} />
      </CardMedia>
      <CardText className="card-description">
        {this.setChar(this.props.article.description,
         120,
         '...'
         )}
         <span className="read-more">
           <a href={this.props.article.url}> read more </a>
          </span>
      </CardText>
      <CardActions className="card-actions">
        <IconButton tooltip={this.state.tooltip} touch
        onClick={this.saveToFavouriteArticles} tooltipPosition="bottom-left">
        {this.state.favIcon}
        </IconButton>>
      </CardActions>
    </Card>
    );
  }
}

ArticleCard.propTypes = {
  article: PropTypes.object
};

export default ArticleCard;
