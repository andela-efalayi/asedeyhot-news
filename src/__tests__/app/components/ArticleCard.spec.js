import React from 'react';
import { shallow } from 'enzyme';
import ArticleCard from '../../../app/components/ArticleCard.jsx';

const articleWithTitle = {
  author: 'BBC News',
  title: 'US submarine arrives in South Korea',
  description: 'It comes amid worries of a North Korean missile',
  url: 'http://www.bbc.co.uk/news/world-asia-39701481',
  urlToImage: 'http://image-url/image.jpg',
  publishedAt: '2017-04-25T02:59:53+00:00'
};
const articleWithNoDescription = {
  author: 'BBC News',
  title: 'US submarine arrives in South Korea',
  description: null,
  url: 'http://www.bbc.co.uk/news/world-asia-39701481',
  urlToImage: 'http://image-url/image.jpg',
  publishedAt: '2017-04-25T02:59:53+00:00'
};

describe('ArticleCard.jsx', () => {
  const wrapper1 = shallow(
    <ArticleCard article={articleWithTitle}/>
  );
  const wrapper2 = shallow(
    <ArticleCard article={articleWithNoDescription}/>
  );
  describe('component', () => {
    it('should match snapshots', () => {
      expect(wrapper1).toMatchSnapshot();
      expect(wrapper2).toMatchSnapshot();
    });
  });

  describe('.card-description', () => {
    it('should have truncated text for long description', () => {
      expect(wrapper1.find('.card-description').node
      .props.children[0].length).toEqual(50);
    });
    it('should show "No text available" if description is null', () => {
      expect(wrapper2.find('.card-description').node
      .props.children[0]).toEqual('No text available');
    });
  });

  describe('.card-actions', () => {
    it('should trigger an onClick event when an icon is clicked', () => {
      wrapper1.find('.card-actions').childAt(0).simulate('click');
      expect(wrapper1.find('.card-actions').childAt(0)
      .node.props.onClick).toBeDefined();
    });
  });
});
