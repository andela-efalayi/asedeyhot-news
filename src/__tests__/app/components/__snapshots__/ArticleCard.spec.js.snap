// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard.jsx component should match snapshots 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Card
    className="card with-image"
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
>
    <CardHeader
        avatar={null}
        className="article-title"
        title="US submarine arrives in South Korea..."
    />
    <CardMedia>
        <img
            src="http://image-url/image.jpg"
        />
    </CardMedia>
    <CardText
        className="card-description"
    >
        It comes amid worries of a North Korean missile...
        <span
            className="read-more"
        >
            <a
                href="http://www.bbc.co.uk/news/world-asia-39701481"
            >
                 read more 
            </a>
        </span>
    </CardText>
    <CardActions
        className="card-actions"
    >
        <IconButton
            disableTouchRipple={false}
            disabled={false}
            iconStyle={Object {}}
            onClick={[Function]}
            tooltip="Add to favourite articles"
            tooltipPosition="bottom-left"
            touch={true}
        >
            <ActionFavoriteBorder />
        </IconButton>
        &gt;
    </CardActions>
</Card>,
  "nodes": Array [
    <Card
      className="card with-image"
      expandable={false}
      expanded={null}
      initiallyExpanded={false}
>
      <CardHeader
            avatar={null}
            className="article-title"
            title="US submarine arrives in South Korea..."
      />
      <CardMedia>
            <img
                  src="http://image-url/image.jpg"
            />
      </CardMedia>
      <CardText
            className="card-description"
      >
            It comes amid worries of a North Korean missile...
            <span
                  className="read-more"
            >
                  <a
                        href="http://www.bbc.co.uk/news/world-asia-39701481"
                  >
                         read more 
                  </a>
            </span>
      </CardText>
      <CardActions
            className="card-actions"
      >
            <IconButton
                  disableTouchRipple={false}
                  disabled={false}
                  iconStyle={Object {}}
                  onClick={[Function]}
                  tooltip="Add to favourite articles"
                  tooltipPosition="bottom-left"
                  touch={true}
            >
                  <ActionFavoriteBorder />
            </IconButton>
            &gt;
      </CardActions>
</Card>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ArticleCard
        article={
                Object {
                        "author": "BBC News",
                        "description": "It comes amid worries of a North Korean missile",
                        "publishedAt": "2017-04-25T02:59:53+00:00",
                        "title": "US submarine arrives in South Korea",
                        "url": "http://www.bbc.co.uk/news/world-asia-39701481",
                        "urlToImage": "http://image-url/image.jpg",
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ArticleCard {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "article": Object {
            "author": "BBC News",
            "description": "It comes amid worries of a North Korean missile",
            "publishedAt": "2017-04-25T02:59:53+00:00",
            "title": "US submarine arrives in South Korea",
            "url": "http://www.bbc.co.uk/news/world-asia-39701481",
            "urlToImage": "http://image-url/image.jpg",
          },
        },
        "refs": Object {},
        "saveToFavouriteArticles": [Function],
        "state": Object {
          "favIcon": <ActionFavoriteBorder />,
          "tooltip": "Add to favourite articles",
          "value": Object {
            "author": "BBC News",
            "description": "It comes amid worries of a North Korean missile",
            "publishedAt": "2017-04-25T02:59:53+00:00",
            "title": "US submarine arrives in South Korea",
            "url": "http://www.bbc.co.uk/news/world-asia-39701481",
            "urlToImage": "http://image-url/image.jpg",
          },
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Card
          className="card with-image"
          expandable={false}
          expanded={null}
          initiallyExpanded={false}
>
          <CardHeader
                    avatar={null}
                    className="article-title"
                    title="US submarine arrives in South Korea..."
          />
          <CardMedia>
                    <img
                              src="http://image-url/image.jpg"
                    />
          </CardMedia>
          <CardText
                    className="card-description"
          >
                    It comes amid worries of a North Korean missile...
                    <span
                              className="read-more"
                    >
                              <a
                                        href="http://www.bbc.co.uk/news/world-asia-39701481"
                              >
                                         read more 
                              </a>
                    </span>
          </CardText>
          <CardActions
                    className="card-actions"
          >
                    <IconButton
                              disableTouchRipple={false}
                              disabled={false}
                              iconStyle={Object {}}
                              onClick={[Function]}
                              tooltip="Add to favourite articles"
                              tooltipPosition="bottom-left"
                              touch={true}
                    >
                              <ActionFavoriteBorder />
                    </IconButton>
                    &gt;
          </CardActions>
</Card>,
        "_debugID": 2,
        "_renderedOutput": <Card
          className="card with-image"
          expandable={false}
          expanded={null}
          initiallyExpanded={false}
>
          <CardHeader
                    avatar={null}
                    className="article-title"
                    title="US submarine arrives in South Korea..."
          />
          <CardMedia>
                    <img
                              src="http://image-url/image.jpg"
                    />
          </CardMedia>
          <CardText
                    className="card-description"
          >
                    It comes amid worries of a North Korean missile...
                    <span
                              className="read-more"
                    >
                              <a
                                        href="http://www.bbc.co.uk/news/world-asia-39701481"
                              >
                                         read more 
                              </a>
                    </span>
          </CardText>
          <CardActions
                    className="card-actions"
          >
                    <IconButton
                              disableTouchRipple={false}
                              disabled={false}
                              iconStyle={Object {}}
                              onClick={[Function]}
                              tooltip="Add to favourite articles"
                              tooltipPosition="bottom-left"
                              touch={true}
                    >
                              <ActionFavoriteBorder />
                    </IconButton>
                    &gt;
          </CardActions>
</Card>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ArticleCard
    article={
        Object {
            "author": "BBC News",
            "description": "It comes amid worries of a North Korean missile",
            "publishedAt": "2017-04-25T02:59:53+00:00",
            "title": "US submarine arrives in South Korea",
            "url": "http://www.bbc.co.uk/news/world-asia-39701481",
            "urlToImage": "http://image-url/image.jpg",
          }
    }
/>,
}
`;

exports[`ArticleCard.jsx component should match snapshots 2`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Card
    className="card with-image"
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
>
    <CardHeader
        avatar={null}
        className="article-title"
        title="US submarine arrives in South Korea..."
    />
    <CardMedia>
        <img
            src="http://image-url/image.jpg"
        />
    </CardMedia>
    <CardText
        className="card-description"
    >
        No text available
        <span
            className="read-more"
        >
            <a
                href="http://www.bbc.co.uk/news/world-asia-39701481"
            >
                 read more 
            </a>
        </span>
    </CardText>
    <CardActions
        className="card-actions"
    >
        <IconButton
            disableTouchRipple={false}
            disabled={false}
            iconStyle={Object {}}
            onClick={[Function]}
            tooltip="Add to favourite articles"
            tooltipPosition="bottom-left"
            touch={true}
        >
            <ActionFavoriteBorder />
        </IconButton>
        &gt;
    </CardActions>
</Card>,
  "nodes": Array [
    <Card
      className="card with-image"
      expandable={false}
      expanded={null}
      initiallyExpanded={false}
>
      <CardHeader
            avatar={null}
            className="article-title"
            title="US submarine arrives in South Korea..."
      />
      <CardMedia>
            <img
                  src="http://image-url/image.jpg"
            />
      </CardMedia>
      <CardText
            className="card-description"
      >
            No text available
            <span
                  className="read-more"
            >
                  <a
                        href="http://www.bbc.co.uk/news/world-asia-39701481"
                  >
                         read more 
                  </a>
            </span>
      </CardText>
      <CardActions
            className="card-actions"
      >
            <IconButton
                  disableTouchRipple={false}
                  disabled={false}
                  iconStyle={Object {}}
                  onClick={[Function]}
                  tooltip="Add to favourite articles"
                  tooltipPosition="bottom-left"
                  touch={true}
            >
                  <ActionFavoriteBorder />
            </IconButton>
            &gt;
      </CardActions>
</Card>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ArticleCard
        article={
                Object {
                        "author": "BBC News",
                        "description": null,
                        "publishedAt": "2017-04-25T02:59:53+00:00",
                        "title": "US submarine arrives in South Korea",
                        "url": "http://www.bbc.co.uk/news/world-asia-39701481",
                        "urlToImage": "http://image-url/image.jpg",
                      }
        }
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ArticleCard {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "article": Object {
            "author": "BBC News",
            "description": null,
            "publishedAt": "2017-04-25T02:59:53+00:00",
            "title": "US submarine arrives in South Korea",
            "url": "http://www.bbc.co.uk/news/world-asia-39701481",
            "urlToImage": "http://image-url/image.jpg",
          },
        },
        "refs": Object {},
        "saveToFavouriteArticles": [Function],
        "state": Object {
          "favIcon": <ActionFavoriteBorder />,
          "tooltip": "Add to favourite articles",
          "value": Object {
            "author": "BBC News",
            "description": null,
            "publishedAt": "2017-04-25T02:59:53+00:00",
            "title": "US submarine arrives in South Korea",
            "url": "http://www.bbc.co.uk/news/world-asia-39701481",
            "urlToImage": "http://image-url/image.jpg",
          },
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Card
          className="card with-image"
          expandable={false}
          expanded={null}
          initiallyExpanded={false}
>
          <CardHeader
                    avatar={null}
                    className="article-title"
                    title="US submarine arrives in South Korea..."
          />
          <CardMedia>
                    <img
                              src="http://image-url/image.jpg"
                    />
          </CardMedia>
          <CardText
                    className="card-description"
          >
                    No text available
                    <span
                              className="read-more"
                    >
                              <a
                                        href="http://www.bbc.co.uk/news/world-asia-39701481"
                              >
                                         read more 
                              </a>
                    </span>
          </CardText>
          <CardActions
                    className="card-actions"
          >
                    <IconButton
                              disableTouchRipple={false}
                              disabled={false}
                              iconStyle={Object {}}
                              onClick={[Function]}
                              tooltip="Add to favourite articles"
                              tooltipPosition="bottom-left"
                              touch={true}
                    >
                              <ActionFavoriteBorder />
                    </IconButton>
                    &gt;
          </CardActions>
</Card>,
        "_debugID": 4,
        "_renderedOutput": <Card
          className="card with-image"
          expandable={false}
          expanded={null}
          initiallyExpanded={false}
>
          <CardHeader
                    avatar={null}
                    className="article-title"
                    title="US submarine arrives in South Korea..."
          />
          <CardMedia>
                    <img
                              src="http://image-url/image.jpg"
                    />
          </CardMedia>
          <CardText
                    className="card-description"
          >
                    No text available
                    <span
                              className="read-more"
                    >
                              <a
                                        href="http://www.bbc.co.uk/news/world-asia-39701481"
                              >
                                         read more 
                              </a>
                    </span>
          </CardText>
          <CardActions
                    className="card-actions"
          >
                    <IconButton
                              disableTouchRipple={false}
                              disabled={false}
                              iconStyle={Object {}}
                              onClick={[Function]}
                              tooltip="Add to favourite articles"
                              tooltipPosition="bottom-left"
                              touch={true}
                    >
                              <ActionFavoriteBorder />
                    </IconButton>
                    &gt;
          </CardActions>
</Card>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ArticleCard
    article={
        Object {
            "author": "BBC News",
            "description": null,
            "publishedAt": "2017-04-25T02:59:53+00:00",
            "title": "US submarine arrives in South Korea",
            "url": "http://www.bbc.co.uk/news/world-asia-39701481",
            "urlToImage": "http://image-url/image.jpg",
          }
    }
/>,
}
`;
